content/typeclasses-syntax.md

```diff
@@ -356,6 +356,256 @@ I believe that a default naming convention for witnesses will be very beneficial
+type implements a type class, or that a value can be used implicitly. We don't need a name for that fact. It's analogous to extends clauses, where we state that a class is a subclass of some other class or trait. We would not think it useful to name an extends clause, it's simply a fact that is stated.
+
+Even for contextual givens, where we do define a value or a function, I tend to prefer to split the definition from the given part. I.e.
+```scala
+  val foo: Foo = ...
+  given Foo = foo
+```
+The given syntax is kind of unwieldy as a way to define values or functions. That was a complaint often voiced when coming from the Scala 2 implicits where this just requires an `implicit` modifier. So my reaction to that is, let's define entities as regular values or functions and inject them as givens separately. But then again, we don't need _another_ name for the given.
+
+Two other arguments for anonymous:
+
+ - Every other language that defines type classes uses anonymous syntax. Somehow, no other language found it necessary to name these instances.
+ - When I define a regular val or def my first concern is: By which name should the defined entity be referenced? That's why the name comes first. For givens this is my least concern. In most cases I don't need a name at all. That's also a reason for the
+ optional suffix names with `as`. Writer as well as readers of given clauses should focus
+on the type that's implemented. That's why the type comes first, and if there is an additional name, we add it at the end with an `as` clause.
+
+The possible variations are presented in the following.
+
+**Current, Anonymous**
+
+```scala
+  // Simple typeclass
+  given Ord[Int] with
+    def compare(x: Int, y: Int) = ...
+
+  // Parameterized typeclass
+  given [A: Ord]: Ord[List[A]] with
+    def compare(x: List[A], y: List[A]) = ...
+
+  // Typeclass with using clause
+  given [A](using Ord[A]): Ord[List[A]] with
+    def compare(x: List[A], y: List[A]) = ...
+
+  // Simple alias
+  given Ord[Int] = IntOrd()
+
+  // Parameterized alias
+  given [A: Ord]: Ord[List[A]] =
+    ListOrd[A]()
+
+  // Alias with using clause
+  given [A](using Ord[A]): Ord[List[A]] =
+    ListOrd[A]
+
+  // Concrete class instance
+  given Context with {}
+
+  // Abstract or deferred given
+  // given Context   // can't be expressed
+
+  // By-name given
+  given [DummySoItsADef]: Context = curCtx
+```
+**Notes:**
+
+ - The `with` is an irregularity with relative to the rest of the language.
+ - The infix `:` also feels strange. To a newcomer, it's not clear what it signifies.
+ - Abstract anonymous givens are not supported.
+ - Concrete class instances need an empty block `with {}`, which is also irregular and might come as a surprise.
+ - By-name givens need an artificial dummy type parameter.
+
+**Proposal, Anonymous**
+
+```scala
+  // Simple typeclass
+  given Ord[Int]:
+    def compare(x: Int, y: Int) = ...
+
+  // Parameterized typeclass
+  given [A: Ord] => Ord[List[A]]:
```

Comment:
I wonder if we could "regularize" this syntax... `[A: Ord] => Ord[List[A]]` looks like a polymorphic function with a context... something like: `[A] => Ord[A] ?=> Ord[List[A]]` So, if we allowed polymorphic context functions with this syntax, `[A: Ord] => Ord[List[A]]` could be a valid type interpreted the same as `[A] => Ord[A] ?=> Ord[List[A]]`.

So, using something close to current syntax:
```scala
given [A: Ord] => Ord[List[A]]:
  [A] => (ord: Ord[A]) ?=> Ord[List[A]]:
    def compare(x: List[A], y: List[A]) = ...
```
now, the given line and the following line is duplicated. If we had a rule that a context function can be written as:
```scala
trait SomeTrait:
  def next(i: Int): Int

val fn: Ctx ?=> SomeTrait:
  //.. we can directly implement SomeTrait without, new, etc...
  next(i: Int) = i + skipSize(using Ctx)
```

Last, we could imagine that if there is a given polymorphic context function, in scope, then it can be used.

If we had the above, this given syntax would fall out of it directly and make the language more regular.

I think the least possible exceptions to syntax helps people learn and read code.

Reply:
Am I understand correctly, your proposal is rewritten this:

```scala
  given [A: Ord] => Ord[List[A]]:
    def compare(x: List[A], y: List[A]) = ...
```
to this:

```scala
  given Ord[A] ?=> Ord[List[A]]:
    def compare(x: List[A], y: List[A]) = ...
```

If that's correct, I really like the idea, because it's just context functions again. `=>` here is a bit awkward imho.